### ISSUE TITLE
ACME plugin generates 403 GoogleCloud API permissions error attempting to update GCS-based Terraform state

### ISSUE THREAD

**Issue by ryan-mf** [paybyphone/terraform-provider-acme#24]
We've been using the paybyphone ACME plugin for Terraform for just shy of a year and it mostly works great....except when we run it in a container. For some reason the plugin works fine for us locally, but fails to be able to write to the Google Cloud Storage-based state when in a container. What's odd is that all other resources native to Terraform work fine, but anything using the ACME plugin fails with a 403 GoogleCloud API error:
```
2017-12-09T00:27:21.283Z [DEBUG] plugin.terraform-provider-acme: 2017/12/09 00:27:21 [INFO][api-ext.prod.***.co] acme: Could not find
2017-12-09T00:27:21.283Z [DEBUG] plugin.terraform-provider-acme: 2017/12/09 00:27:21 [INFO][api-ext.prod.***.co] acme: Trying to solv
2017/12/09 00:27:21 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* acme_certificate.apiext_ssl_certificate: Error presenting token: GoogleCloud API call failed: googleapi: Error 403: Insufficient Permission,
2017/12/09 00:27:21 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* acme_certificate.apiext_ssl_certificate: Error presenting token: GoogleCloud API call failed: googleapi: Error 403: Insufficient Permission,
2017/12/09 00:27:21 [TRACE] [walkApply] Exiting eval tree: acme_certificate.apiext_ssl_certificate
2017/12/09 00:27:21 [INFO] Writing tf-state-prod/terraform/terraform.tfstate
Error applying plan:

2 error(s) occurred:

* acme_certificate.apiint_ssl_certificate: 1 error(s) occurred:

* acme_certificate.apiint_ssl_certificate: Error presenting token: GoogleCloud API call failed: googleapi: Error 403: Insufficient Permission,
2017/12/09 00:27:21 [DEBUG] plugin: waiting for all plugin processes to complete...
* acme_certificate.apiext_ssl_certificate: 1 error(s) occurred:

* acme_certificate.apiext_ssl_certificate: Error presenting token: GoogleCloud API call failed: googleapi: Error 403: Insufficient Permission,

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```
Because Terraform is able to write to the state just fine for all other native resources, and because the ACME plugin isn't giving debug output, I can't tell what user it's trying to use (maybe different from what Terraform is trying to use natively?) to write to the state which results in a 403. In any case, I've ensured that both the service account I've told Terraform to use, *and* the service account that the GKE cluster gets by default, have read/write access to the storage bucket that contains the Terraform state.

Why would Terraform be able to write to the state, but ACME plugin not?

**Comment by scream314** [comment]
@ryan-mf 
We are using the plugin with AWS, but we have to specify `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY` and `AWS_REGION` in `resource.acme_certificate.my_cert`.
Are you sure the correct credentials are passed resource? Maybe you are using environment variables outside the container and these are not correctly set inside?

**Comment by ryan-mf** [comment]
Hi @scream314,

So trying to understand this. Even though it works locally, but not in a container, the likely answer is that it's never actually used the service account I told Terraform to use (and which the other Terraform resources all use), but instead the acme module is likely using my own gcloud authentication or something? The container uses gcloud to authenticate with a service account on startup so I would think that would work too, but apparently not.

As a result, I'm not understanding what VARs I'd have to add to the acme_certificate resource. We're using Google Cloud, not AWS. The root user in the container has a valid ~/.config/gcloud directory with a valid login setup.

**Comment by ryan-mf** [comment]
I ended up figuring this out with your pointer, @scream314, thanks! Indeed, the ACME module wasn't trying to use the auth info that Terraform was set to use (and which works with all other of our Terraform resources) - instead it was looking for the ENVVAR GOOGLE_APPLICATION_CREDENTIALS. I don't have that set locally and it still works, so it must be using some other credentials discovery mechanism. Once I set GOOGLE_APPLICATION_CREDENTIALS in the container, it all started to work.

Thanks!
